### PR TITLE
RuntimeBroker is now default broker for non-UWP

### PR DESCRIPTION
**Pull request by SameerK-MSFT**
Fixes #
Changes Old WAM to new WAM close to https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3447

**Changes proposed in this request**
- WithWindowsBroker now invokes RuntimeBroker
- WithWindowsDesktop invokes the old broker
- 

**Testing**
Manual tests with apps
Unit + integration tests 

**Performance impact**
NA. Desktop package takes dependency on Broker. It will cause size increase.

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client.Desktop/DesktopExtensions.cs [hunk 1]
```diff
@@ -19,6 +20,7 @@ public static class DesktopExtensions
         /// - WebView2 embedded web view, based on Microsoft Edge - https://aka.ms/msal-net-webview2
         /// </summary>
         /// <remarks>These extensions live in a separate package to avoid adding dependencies to MSAL</remarks>
+        [Obsolete("This method has been deprecated. Please use equivalent method WithWindowsDesktopFeatures")]
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Desktop/DesktopExtensions.cs:23]:
```suggestion
        [Obsolete("This method has been deprecated. Use WithWindowsDesktopFeatures. For Windows Broker support only, use WithWindowsBroker API from Microsoft.Identity.Client.Broker package.")]
```

File: src/client/Microsoft.Identity.Client.Desktop/DesktopExtensions.cs [hunk 2]
```diff
@@ -27,6 +29,23 @@ public static PublicClientApplicationBuilder WithDesktopFeatures(this PublicClie
             return builder;
         }
 
+        /// <summary>
+        /// Adds enhanced support for desktop applications, e.g. CLI, WinForms, WPF apps.
+        /// 
+        /// Support added is around: 
+        /// 
+        /// - Windows Authentication Manager (WAM) broker, the recommended authentication mechanism on Windows 10 - https://aka.ms/msal-net-wam
+        /// - WebView2 embedded web view, based on Microsoft Edge - https://aka.ms/msal-net-webview2
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Desktop/DesktopExtensions.cs:38]:
```suggestion
        /// - Embedded web view. AAD applications use the older WebBrowser control. B2C applications use WebView2, an embedded browser based on Microsoft Edge - https://aka.ms/msal-net-webview2
```

**Review by bgavrilMS — COMMENTED** [review]
LGTM

**Review by bgavrilMS — APPROVED** [review]

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs [hunk 3]
```diff
@@ -46,7 +46,7 @@ public async Task WamSilentAuthUserInteractionRequiredAsync()
                .Create("04f0c124-f2bc-4f59-8241-bf6df9866bbd")
                .WithAuthority("https://login.microsoftonline.com/organizations");
 
-            IPublicClientApplication pca = pcaBuilder.WithBrokerPreview().Build();
+            IPublicClientApplication pca = pcaBuilder.WithWindowsBroker().Build();
```
Comment (gladjohn) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:49]: Can you also add a integration test for the Desktop Broker calling into Runtime Broker?

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs [hunk 4]
```diff
@@ -110,7 +110,7 @@ public async Task WamUsernamePasswordRequestAsync()
                .WithParentActivityOrWindow(windowHandleProvider)
                .WithAuthority(labResponse.Lab.Authority, "organizations")
                .WithLogging(testLogger)
-               .WithBrokerPreview().Build();
+               .WithWindowsBroker().Build();
```
Comment (gladjohn) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:113]: You should be able to add the same test for Desktop Broker as well

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs [hunk 5]
```diff
@@ -71,7 +71,7 @@ public void WamOnWin10()
                .Create("d3adb33f-c0de-ed0c-c0de-deadb33fc0d3")
                .WithAuthority(TestConstants.AuthorityTenant);
 
-            pcaBuilder = pcaBuilder.WithBrokerPreview();
+            pcaBuilder = pcaBuilder.WithWindowsBroker();
```
Comment (gladjohn) [inline tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:74]: Why is the test failing?

File: tests/devapps/NetCoreTestApp/Program.cs [hunk 6]
```diff
@@ -275,6 +280,26 @@ 1. IWA
                             }
 
                             break;
+
+                        case 12: // acquire token interactive with WamBroker
+                            {
+                                var optionsbroker = new SystemWebViewOptions()
+                                {
+                                    //BrowserRedirectSuccess = new Uri("https://www.bing.com?q=why+is+42+the+meaning+of+life")
```
Comment (gladjohn) [inline tests/devapps/NetCoreTestApp/Program.cs:288]: ?

**Review by gladjohn — APPROVED** [review]
where applicable please add integration test for Desktop Broker and it should call into the new broker and work. like ROPC / Scope check etc.

File: src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs [hunk 7]
```diff
@@ -23,6 +24,26 @@ public static class BrokerExtension
         /// If a broker does not exist or cannot be used, MSAL will fallback to a browser.
         /// Make sure browser auth is enabled (e.g. if using system browser, register the "http://localhost" redirect URI, etc.)
         /// </remarks>
+        // The name to change later
+        public static PublicClientApplicationBuilder WithWindowsBroker(this PublicClientApplicationBuilder builder, bool enableBroker = true)
+        {
+            builder.Config.IsBrokerEnabled = enableBroker;
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs:30]: Add logging info

File: src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs [hunk 8]
```diff
@@ -23,6 +24,26 @@ public static class BrokerExtension
         /// If a broker does not exist or cannot be used, MSAL will fallback to a browser.
         /// Make sure browser auth is enabled (e.g. if using system browser, register the "http://localhost" redirect URI, etc.)
         /// </remarks>
+        // The name to change later
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs:27]: Good catch.

File: src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs [hunk 9]
```diff
@@ -23,6 +24,26 @@ public static class BrokerExtension
...
+        /// For details about Windows broker, see https://aka.ms/msal-net-wam
+        /// </summary>
+        /// <remarks>
+        /// No broker integration exists on Mac and Linux yet.
+        /// Windows broker does not work on Win 8, Win Server 2016 and lower.
+        /// This implementation is not supported on UWP, use <c>WithBroker()</c> from Microsoft.Identity.Client package instead.
+        /// If a broker does not exist or cannot be used, MSAL will fallback to a browser.
+        /// Make sure browser auth is enabled (e.g. if using system browser, register the "http://localhost" redirect URI, etc.)
+        /// </remarks>
+        [Obsolete("Thanks for trying the preview. This method has been deperecated. Please use WithWindowsBroker now. And WithExperimentalFeatures need not be used.")]
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs]: Fixed in another way. Will update

File: src/client/Microsoft.Identity.Client.Desktop/WamExtension.cs [hunk 10]
```diff
@@ -18,11 +19,11 @@ public static class WamExtension
         /// </summary>
         public static PublicClientApplicationBuilder WithWindowsBroker(this PublicClientApplicationBuilder builder, bool enableBroker = true)
         {
-            builder.Config.IsBrokerEnabled = enableBroker;
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Desktop/WamExtension.cs:21]: Removed in the latest

File: tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs (hunk 5 above)
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Unit/BrokerTests/RuntimeBrokerTests.cs:74]: It fails with error unable to load msal_runtime.

File: tests/devapps/NetCoreTestApp/Program.cs (hunk 6 above)
Comment (SameerK-MSFT) [reply tests/devapps/NetCoreTestApp/Program.cs:288]: Copy paste from #238. https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/f0ec8124f6d927bea3a3f1040197e26c74066a90/tests/devapps/NetCoreTestApp/Program.cs#L238
Will remove from both the places.

File: src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs:27]: Yep. Leftover

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs (hunk 3 above)
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:49]: That is static method call. So, I would avoid writing test for it. If you have suggestions, please let me know.

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs (hunk 4 above)
Comment (SameerK-MSFT) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs:113]: There are pre-existing tests for WamBroker. I do not think any new are needed.

File: tests/devapps/NetCoreTestApp/Program.cs [hunk 11]
```diff
@@ -66,14 +66,18 @@ private static string GetAuthority()
...
                             .WithAuthority(GetAuthority())
-                            .WithLogging(Log, LogLevel.Verbose, true)
-                            .WithExperimentalFeatures()
-                            .WithDesktopFeatures();
+                            .WithLogging(Log, LogLevel.Verbose, true);
+
+            if(withWamBroker)
+            {
+                pcaBuilder.WithWindowsDesktopFeatures()
+                            .WithBroker();
```
Comment (pmaytak) [inline tests/devapps/NetCoreTestApp/Program.cs:79]: Why two calls?

File: tests/devapps/NetCoreTestApp/Program.cs [hunk 12]
```diff
@@ -275,6 +279,25 @@ 1. IWA
                             }
 
                             break;
+
+                        case 12: // acquire token interactive with WamBroker
```
Comment (pmaytak) [inline tests/devapps/NetCoreTestApp/Program.cs:283]: As a team we should clarify which broker we're referring to. I propose to explicitly state 'legacy' for anything related to the old broker.

File: tests/devapps/NetCoreTestApp/Program.cs [hunk 13]
```diff
@@ -275,6 +279,25 @@ 1. IWA
...
+                            {
+                                var optionsbroker = new SystemWebViewOptions()
```
Comment (pmaytak) [inline tests/devapps/NetCoreTestApp/Program.cs:285]: Why are these added? They don't affect the broker?

File: tests/devapps/WAM/NetCoreWinFormsWam/Form1.cs [hunk 14]
```diff
@@ -120,14 +120,14 @@ private IPublicClientApplication CreatePca(AuthMethod? authMethod)
                     builder = ToggleOldBroker(builder, true);
                     break;
                 case AuthMethod.WAMRuntime:
-                    builder = builder.WithBrokerPreview().WithExperimentalFeatures();
+                    builder = BrokerExtension.WithWindowsBroker(builder);
```
Comment (pmaytak) [inline tests/devapps/WAM/NetCoreWinFormsWam/Form1.cs:123]: Why are the extension classes explicitly called here and not through the `builder`?

File: tests/Microsoft.Identity.Test.Unit/AppConfigTests/PublicClientApplicationBuilderTests.cs [hunk 15]
```diff
@@ -649,7 +649,7 @@ public void IsBrokerAvailable_OldDotNet()
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/AppConfigTests/PublicClientApplicationBuilderTests.cs:649]: Do we have full test coverage to test that correct broker implementation is used for all platforms? Do we test net6_win anywhere? We can add more tests here (or wherever is more appropriate).

**Review by pmaytak — COMMENTED** [review]
[WithBroker](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/3d20374e4bd9ed621119b917c51e4c1eb68d3032/src/client/Microsoft.Identity.Client/AppConfig/PublicClientApplicationBuilder.cs#L155) needs to be updated:
- NET461, NET_CORE error messages need to reflect the broker change.
- net6_win apps use this method, and still end up using the old broker.

File: src/client/Microsoft.Identity.Client.Desktop/DesktopExtensions.cs [hunk 16]
```diff
@@ -27,6 +29,23 @@ public static PublicClientApplicationBuilder WithDesktopFeatures(this PublicClie
...
+        /// <summary>
+        /// Adds enhanced support for desktop applications, e.g. CLI, WinForms, WPF apps.
+        /// 
+        /// Support added is around: 
+        /// 
+        /// - Windows Authentication Manager (WAM) broker, the recommended authentication mechanism on Windows 10 - https://aka.ms/msal-net-wam
+        /// - Embedded web view. AAD applications use the older WebBrowser control. B2C applications use WebView2, an embedded browser based on Microsoft Edge - https://aka.ms/msal-net-webview2
+        /// </summary>
+        /// <remarks>These extensions live in a separate package to avoid adding dependencies to MSAL</remarks>
+        public static PublicClientApplicationBuilder WithWindowsDesktopFeatures(this PublicClientApplicationBuilder builder)
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Desktop/DesktopExtensions.cs:41]: The one above is effectively renamed to have platform specific name for Legacy availability. At the same time, message encourages developers to use the new API.

File: src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs [hunk 17]
```diff
@@ -23,6 +24,25 @@ public static class BrokerExtension
         /// If a broker does not exist or cannot be used, MSAL will fallback to a browser.
         /// Make sure browser auth is enabled (e.g. if using system browser, register the "http://localhost" redirect URI, etc.)
         /// </remarks>
+        public static PublicClientApplicationBuilder WithWindowsBroker(this PublicClientApplicationBuilder builder, bool enableBroker = true)
```
[the remark this answers is not in the document]
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs:27]: Yeah, I think we want to add `WithMacBroker` and `WithLinuxBroker` later. The danger of putting them under a single API is that you'll enable the broker, it'll default to browser on Mac, then we add support for Mac, developer picks it up without testing, and that'll break them because redirect URI for Mac isn't setup.

File: tests/devapps/NetCoreTestApp/Program.cs (hunk 12 above)
Comment (bgavrilMS) [reply tests/devapps/NetCoreTestApp/Program.cs:283]: Hmm, there really should not be a "legacy" broker. We only use the new broker. Only on UWP we use the "legacy" broker because the new broker does not support UWP.

File: src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Broker/BrokerExtension.cs:27]: +1 to Bogdan. I provided the same answer in the API design doc.

File: src/client/Microsoft.Identity.Client.Desktop/WamExtension.cs [hunk 18]
```diff
@@ -18,11 +19,11 @@ public static class WamExtension
...
-            AddSupportForWam(builder);
-            return builder;
+            builder.Config.IdentityLogger?.Log(new IdentityModel.Abstractions.LogEntry() { EventLogLevel = IdentityModel.Abstractions.EventLogLevel.Informational, Message = "Desktop WAM Broker extension calling RuntimeBroker extension" });
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Desktop/WamExtension.cs:22]: The method does not have logger supplied to it. Also, this is logged during initialization, whereas the actual broker is invoked elsewhere. This should be OK IMO.

File: src/client/Microsoft.Identity.Client.Desktop/WamExtension.cs [hunk 19]
```diff
@@ -18,11 +19,11 @@ public static class WamExtension
...
+            return BrokerExtension.WithWindowsBroker(builder, enableBroker);
         }
 
+        // this is for legacy WAM
         internal static void AddSupportForWam(PublicClientApplicationBuilder builder)
```
[the remark this answers is not in the document]
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client.Desktop/WamExtension.cs:27]: Done

File: tests/devapps/NetCoreTestApp/Program.cs (hunk 11 above)
Comment (SameerK-MSFT) [reply tests/devapps/NetCoreTestApp/Program.cs:79]: WithBroker is required for config. This was there in different file earlier too. [Ref:]https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/da70e8a1bae4ab8ddb1ce659e7ed4c5167aa0334/tests/devapps/NetFxConsoleTestApp/Program.cs#L142)

File: tests/devapps/NetCoreTestApp/Program.cs (hunk 12 above)
Comment (SameerK-MSFT) [reply tests/devapps/NetCoreTestApp/Program.cs:283]: Case 12 is for usign the old/legacy WamBroker

File: tests/devapps/NetCoreTestApp/Program.cs (hunk 13 above)
Comment (SameerK-MSFT) [reply tests/devapps/NetCoreTestApp/Program.cs:285]: Case 12 is for using the old/legacy WamBroker

File: tests/devapps/WAM/NetCoreWinFormsWam/Form1.cs (hunk 14 above)
Comment (SameerK-MSFT) [reply tests/devapps/WAM/NetCoreWinFormsWam/Form1.cs:123]: The intention is to specifically test BrokerExtension.
